csrc/lower_warp_reduce.cpp

```diff
@@ -148,6 +148,10 @@ class FuseBroadcastWithWarpReduce : private kir::IrVisitor {
  public:
   static std::vector<Expr*> fuse(const std::vector<Expr*>& exprs) {
     FuseBroadcastWithWarpReduce fuse_broadcast_map(exprs);
+    if (fuse_broadcast_map.val_replacement_map_.empty()) {
```

Comment:
Is the issue the broadcast elimination pass incorrectly eliminates broadcast ops used for gather ops? If so, what happens if the warp fusion indeed happens?

Reply:
`EliminateDeadBroadcastAndAllocate` looks for any local memory `Allocate`s whose buffers are broadcast outputs and which are not the inputs to an expression. It then marks any of these that are inputs to an expression in the kernel expression tree as "live", and removes any of the others. The thing is that a gather op actually looks like a `Set` op where the only input is the tensor being indexed, and the index expression itself is an attribute instead of an input. A more complete fix may be to check attributes in the Expr list as well, and look at those recursively, but I thought that may be too intrusive.

Reply:
I also considered special casing for set ops. But I'm currently just looking at what is the actual difference between this pass and the earlier grid reduce fusion pass. In this pass, we check that the scopes are the same between the reduction and broadcast. In a graph-based pass, we could check that they are inlined at the same level instead I think? Otherwise, there is an assumption for fused warp reduce that the input is in register/local mem; I'm not sure if that also holds for grid reduce.

Reply:
Hmm, maybe I'm missing something, but if we update this part: 

https://github.com/NVIDIA/Fuser/blob/main/csrc/lower_warp_reduce.cpp#L97-L103

to include the use by `gather`, isn't the problem gone?

Reply:
Yeah I think we should be able to detect it, but I think it needs a bit of traversal. The reason is that in this case, the missing allocation is only used inside a somewhat deep indexing expression. That expression is a `TensorIndex` that's the input to a `LoadStoreOp` that we could detect there where you pointed. However, we'd then need to traverse to find the actual tensor we need to mark as live (by finding any root TVs of the `index()` I suppose). I can take a crack at it.

Reply:
Latest push removes the empty check and adds all TV inputs of TensorIndex index() as live TVs. Note that it doesn't recurse further, since IterVisitor doesn't currently pass through `.index()` of a `TensorIndex` either. If this is the only case, we could add that level to IterVisitor. This is where Attributes on Vals might be nice, but Attributes are themselves Vals so it may not be so easy to add.